File: checkerGeneralArchitecture.java
public enum boardSize {small, medium, large}
public enum gameStatus {playing, win, draw}
public enum playerType {human, computer}
public enum pieceColor {black, white}
public enum pieceType {regular, king}
public enum moveType {regular, jump}
public enum moveStatus {legal, logicError} // syntax error should be checked in the UI part
public enum listOfMessages {} // all possible ui messages 

public class GameManager 
{
	Board board;
	Player player1;
	Player player2;
	Move[] moveHistory;
	Player currentPlayer;

	public GameManager()
	{
		// default constructor
	}

	public void Start()
	{
		// the function that runs the entire game
	}

}

public class Board 
{
	Piece[][] board;
	boardSize size;

	public Board() 
	{
		// default constructor
	}

	public Board(boardSize size)
	{
		// create a new board of specific size, init it with pieces
	}

	private void initBoard() 
	{
		// initialized the board with the right amount of pieces
	}

	public void MovePiece(Move move, out moveStatus)
	{
		// if the move is legit, move the piece
	}

	private moveStatus checkMoveLegalality(Move move)
	{
		// check if the move is legal
	} 

	public Move[] GetPossibleMoves(Player currentPlayer, Move lastMove)
	{
		// returns the list of all possible moves for a specific player
		// if the last move was a jump, first check if another jump is possible
	}

	public Piece[][] GetBoard()
	{
		// returns the board matrix
	}

	public gameStatus GetGameStatus() 
	{
		// check the current game status, without returning the winner and points
	}

	public gameStatus GetGameStatus(out Player winner, out int points) 
	{
		// check the current game status
		// if there is a win or a draw, return the winner and the amount of points he got
	}

}

public class CheckersConsolUI
{

	PUBLIC STATIC FINAL int MAX_NAME_SIZE = 20;
	public static void printBoard(Board board)
	{
		// print the board
	}

	public static String getUserInput() 
	{
		 // Get an input from the user
	}

	public static void clearScreen()
	{
		// call the dll function that clears the screen
	}

	private static boolean checkNameSize(String name)
	{
		return (name.length() <= MAX_NAME_SIZE)
	}

	private static boolean parseMove(String userInput, move move)
	{
		// parse the user input into a move type or return false
	}

	public void printMessage(listOfMessages) 
	{
		// prints some message to the screen
	}

}

public interface Player 
{
	playerType type;
	pieceColor color;
	Move[] moveHistory;
	int points;

	public abstract move GetMove() 

	public void ChangePoints(int amount)
	{
		points += decrease;
		// When losing a piece, decrease by 1
		// When losing a king, decrease by 4
		// When getting a king, increase by 3 (make sure this is true)
	} 

	public int GetPoints() 
	{
		return points;
	}

	public void ClearMoveHistory()
	{
		// clears the move history, in order to start a new game
	}
}

public class Human extends Player 
{
	String name;

	public Human(pieceColor color) 
	{
		type = playerType.human;
		this.color = color;
	}


	public move GetMove() 
	{
		// ask the UI for a move
	}

	public void SetName() 
	{
		// sets the players name
	}

}

public class Computer extends Player 
{
	public Computer(pieceColor color) 
	{
		type = playerType.computer;
		this.color = color;
	}

	public move GetMove()
	{
		// ask the board for a list of moves, return a random one
	}

}

public class Move 
{
	Position begin;
	Position End;
	int player;
	moveType type;

}

public class Position 
{
	int row;
	int col;
}

public class Piece
{
	pieceColor color;
	pieceType type;
}

File: checkerGeneralArchitecture.cs
// add e_ to all enums
public enum boardSize {small, medium, large}
public enum gameStatus {playing, win, draw}
public enum playerType {human, computer}
public enum pieceColor {black, white}
public enum pieceType {regular, king}
public enum moveType {regular, jump}
public enum moveStatus {legal, logicError} // syntax error should be checked in the UI part
public enum listOfMessages {} // all possible ui messages 

public class GameManager 
{
	Board m_Board;
	Player? m_Player1;
	Player? m_Player2;
	Move?[] m_MoveHistory;
	Player? m_CurrentPlayer;

	public GameManager()
	{
		// default constructor
	}

	public void Start()
	{
		// the function that runs the entire game
	}

	private moveStatus checkMoveLegalality(Move i_Move)
	{
		// check if the move is legal
	} 

	private Move getRandomMove(Player i_Player)
	{
		// returns a move for a specific player, from all possible moves 
	}

	private void changeActivePlayer()
	{
		// changes the active player to the other one
	}

}

protected class Board 
{
	Piece?[][] m_Board;
	boardSize m_Size;

	protected Board() 
	{
		// default constructor
	}

	protected Board(boardSize i_size)
	{
		// create a new board of specific size, init it with pieces
	}

	private void initBoard() 
	{
		// initialized the board with the right amount of pieces
		// init null where there are no objects
	}

	protected void MovePiece(Move i_Move, out moveStatus) // naming convention for out variables? 
	{
		// if the move is legit, move the piece
	}

	protected Move[] GetPossibleMoves(Player i_CurrentPlayer, Move i_LastMove)
	{
		// returns the list of all possible moves for a specific player
		// if the last move was a jump, first check if another jump is possible
	}

	protected Piece[][] GetBoard()
	{
		// returns the board matrix
	}

	protected gameStatus GetGameStatus() 
	{
		// check the current game status, without returning the winner and points
	}

	protected gameStatus GetGameStatus(out Player winner, out int points) 
	{
		// check the current game status
		// if there is a win or a draw, return the winner and the amount of points he got
	}

}

protected class CheckersConsolUI
{

	PUBLIC STATIC FINAL int MAX_NAME_SIZE = 20;

	protected static void PrintBoard(Board i_Board)
	{
		// print the board
	}

	protected static String GetUserInput() 
	{
		 // Get an input from the user
	}

	protected static void ClearScreen()
	{
		// call the dll function that clears the screen
	}

	private static bool checkName(String i_Name)
	{
		return (i_Name.Length <= MAX_NAME_SIZE)
	}

	private static bool parseMove(String i_UserInput, move i_Move)
	{
		// parse the user input into a move type or return false
	}

	protected void PrintMessage(listOfMessages i_Message) 
	{
		// prints some message to the screen
	}

}

protected class Player 
{
	private String? m_Name;
	private playerType m_Type;
	private pieceColor m_Color;
	private Move?[] m_MoveHistory;
	private int m_Points = 0;

	protected Name 
	{
		get
		{
			return m_Name;
		}
		set
		{
			m_Name = value; 
		}
	}

	protected Points 
	{
		get
		{
			return m_Points; 
		}

		set
		{
			m_Points = value;
		}
	}

	protected Type
	{
		get 
		{
			return m_Type;
		}
		set
		{
			m_Type = value;
		}
	}

	protected Color
	{
		get
		{
			return m_Color;
		}
		set
		{
			m_Color = value;
		}

	}

	protected void ClearMoveHistory()
	{
		// clears the move history, in order to start a new game
	}

	protected void AddMove()
	{
		// add a move to the players move history
	}
}


protected struct Move 
{
	private Position m_Begin;
	private Position m_End;
	private Player m_Player;
	private moveType m_Type;

	protected Move(Position i_Begin, Position i_End, Player i_Player)
	{
		// new move constructor 
	}

	// get begin 
	// get end
	// get player
	// get/set move type

}

protected struct Position 
{
	private int m_Row;
	private int m_Col;

	protected Position(int i_Row, int i_Col)
	{
		// new position constructor
	}
}

protected Struct Piece
{
	private pieceColor m_Color;
	private pieceType m_Type;

	protected Piece(pieceColor i_Color, pieceType i_Type) 
	{
		// new piece constructor
	}

	protected void SetKing() 
	{
		// change piece type from regular to king
	}
}

